### PR TITLE
Add support for assertions

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we add a new `assertions` property for operations. When
this property is present and assertions are enabled with `--enable-assertions`,
Rally checks all assertions and raises an error if the check fails.

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
This has been caused by a transient failure:

```
[2021-01-25T12:53:07,103][ERROR][o.e.b.Bootstrap          ] [rally-node-0] Exception
org.elasticsearch.http.BindHttpException: Failed to bind to [39200]
[...]
Caused by: java.net.BindException: Address already in use
```

@elasticmachine test this please

File: docs/track.rst [hunk 1]
```diff
@@ -1619,9 +1663,18 @@ composite
...
 
-* ``requests`` (mandatory): A list that specifies the request streams to execute. Each stream consists of operations of either type ``raw-request`` to send requests to Elasticsearch or ``sleep`` to simulate client processing time. Streams execute concurrently, operations within a stream sequentially. It is possible to nest streams. See below for specific examples.
+* ``requests`` (mandatory): A list that specifies the request streams to execute. Streams execute concurrently, operations within a stream sequentially. It is possible to nest streams. See below for specific examples.
 * ``max-connections`` (optional: defaults to unbounded): The maximum number of concurrent connections per client executing this composite operation. By default, the operation itself does not restrict the number of connections but is bound to Rally's network connection limit. Therefore raise the number of available network connections appropriately (see :ref:`command line reference <clr_client_options>`).
 
+The ``composite`` operation only supports the following operation-types:
+
+* ``raw-request``
+* ``sleep``
+* ``search``
```
Comment (gingerwizard) [inline docs/track.rst:1673]: Adding search to composite 👍

File: docs/track.rst [hunk 2]
```diff
@@ -566,10 +566,54 @@ Each operation consists of the following properties:
...
+* ``assertions`` (optional, defaults to ``None``): A list of assertions that should be checked. See below for more details.
 * ``request-timeout`` (optional, defaults to ``None``): The client-side timeout for this operation.  Represented as a floating-point number in seconds, e.g. ``1.5``.
 * ``headers`` (optional, defaults to ``None``): A dictionary of key-value pairs to pass as headers in the operation.
 * ``opaque-id`` (optional, defaults to ``None`` [unused]): A special ID set as the value of ``x-opaque-id`` in the client headers of the operation.  Overrides existing ``x-opaque-id`` entries in ``headers`` (case-insensitive).
 
+**Assertions**
+
+Use assertions for sanity checks, e.g. to ensure a query returns results. Assertions need to be defined with the following properties. All of them are mandatory:
+
+* ``property``: A dot-delimited path to the meta-data field to be checked. Only meta-data fields that are returned by an operation are supported.
```
Comment (gingerwizard) [inline docs/track.rst:578]: Maybe im missing something but i cant find a list of meta fields without looking at the runner. Is this in the docs somewhere?

File: esrally/driver/runner.py [hunk 3]
```diff
@@ -2042,15 +2124,26 @@ async def __call__(self, es, params):
...
+        stats = {}
+        for search_id, search in async_search_ids(searches):
             response = await es.async_search.get(id=search_id,
                                                  params=request_params)
-            success = success and not response["is_running"]
+            is_running = response["is_running"]
+            success = success and not is_running
+            if not is_running:
+                stats[search] = {
+                    "hits": response["response"]["hits"]["total"]["value"],
```
Comment (gingerwizard) [inline esrally/driver/runner.py:2135]: Do we document these as the "meta-fields" available for assertions?

File: esrally/track/params.py [hunk 4]
```diff
@@ -515,24 +517,12 @@ def __init__(self, track, params, **kwargs):
             self.query_params["pages"] = pages
         if results_per_page:
             self.query_params["results-per-page"] = results_per_page
+        if "assertions" in params:
+            self.query_params["assertions"] = params["assertions"]
```
Comment (gingerwizard) [inline esrally/track/params.py:521]: Should we error if `detailed-results` is False and `assertions` is True?

File: esrally/driver/runner.py [hunk 5]
```diff
@@ -311,6 +324,69 @@ async def __aexit__(self, exc_type, exc_val, exc_tb):
...
+        self.predicates = {
+            ">": self.greater_than,
+            ">=": self.greater_than_or_equal,
+            "<": self.smaller_than,
+            "<=": self.smaller_than_or_equal,
+            "==": self.equal,
+        }
+
+    def greater_than(self, expected, actual):
+        return actual > expected
```
Comment (gingerwizard) [inline esrally/driver/runner.py:341]: I notice there are no type restrictions here? i.e. we could in theory compare strings. Is this intentional and left to the user to ensure valid types are being compared?

**Review by gingerwizard — APPROVED** [review]
A few clarifications if possible but more for my understanding re typing and meta fields

File: docs/track.rst (hunk 2 above)
Comment (danielmitterdorfer) [reply docs/track.rst:578]: You did not miss anything; this is not documented at the moment. I propose we do this in a follow-up PR.

File: esrally/driver/runner.py (hunk 3 above)
Comment (danielmitterdorfer) [reply esrally/driver/runner.py:2135]: As indicated above: No, it's currently undocumented.

File: esrally/track/params.py (hunk 4 above)
Comment (danielmitterdorfer) [reply esrally/track/params.py:521]: Good idea. :)

File: esrally/driver/runner.py (hunk 5 above)
Comment (danielmitterdorfer) [reply esrally/driver/runner.py:341]: Yes, it's intentional that we're not too restrictive here.

File: esrally/track/params.py (hunk 4 above)
Comment (danielmitterdorfer) [reply esrally/track/params.py:526]: I've pushed https://github.com/elastic/rally/pull/1157/commits/8286f9fd17eb04666bb370200e81d97f4b924e24 to address your proposal. Can you please have another look @gingerwizard?

File: esrally/driver/runner.py (hunk 5 above)
Comment (gingerwizard) [reply esrally/driver/runner.py:341]: 👍 maybe we could warn on type comparisons but probably not worth the cost.

**Pull request merged**